publisher/publisher_test.go

```diff
@@ -432,3 +433,53 @@ func TestLogErrorBody(t *testing.T) {
 	test.AssertError(t, err, "SubmitToSingleCTWithResult didn't fail")
 	test.AssertEquals(t, len(log.GetAllMatching("well this isn't good now is it")), 1)
 }
+
+func TestProbeLogs(t *testing.T) {
+	pub, _, k := setup(t)
+
+	srvA := logSrv(k)
+	defer srvA.Close()
+	portA, err := getPort(srvA.URL)
+	test.AssertNotError(t, err, "Failed to get test server port")
+	srvB := errorBodyLogSrv()
+	defer srvB.Close()
+	portB, err := getPort(srvB.URL)
+	test.AssertNotError(t, err, "Failed to get test server port")
+
+	k1, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	test.AssertNotError(t, err, "ecdsa.GenerateKey() failed for k1")
+	der1, err := x509.MarshalPKIXPublicKey(&k1.PublicKey)
+	test.AssertNotError(t, err, "x509.MarshalPKIXPublicKey(der1) failed")
+	k1b64 := base64.StdEncoding.EncodeToString(der1)
+	_, err = pub.ctLogsCache.AddLog(fmt.Sprintf("http://localhost:%d", portA), k1b64, pub.log)
+	test.AssertNotError(t, err, "Failed to add log to logCache")
+	k2, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	test.AssertNotError(t, err, "ecdsa.GenerateKey() failed for k2")
+	der2, err := x509.MarshalPKIXPublicKey(&k2.PublicKey)
+	test.AssertNotError(t, err, "x509.MarshalPKIXPublicKey(der2) failed")
+	k2b64 := base64.StdEncoding.EncodeToString(der2)
+	_, err = pub.ctLogsCache.AddLog(fmt.Sprintf("http://localhost:%d", portB), k2b64, pub.log)
+	test.AssertNotError(t, err, "Failed to add log to logCache")
+	k3, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	test.AssertNotError(t, err, "ecdsa.GenerateKey() failed for k3")
+	der3, err := x509.MarshalPKIXPublicKey(&k3.PublicKey)
+	test.AssertNotError(t, err, "x509.MarshalPKIXPublicKey(der3) failed")
+	k3b64 := base64.StdEncoding.EncodeToString(der3)
+	_, err = pub.ctLogsCache.AddLog("http://blackhole:9999", k3b64, pub.log)
+	test.AssertNotError(t, err, "Failed to add log to logCache")
```

Comment:
This looks like a place where a small inner function to generate a key, marshal it, encode it, and add it to the cache would clean up some duplication.